Add namespace and autoloading

cliffordp opened the pull request:
https://central.tri.be/issues/116489

cliffordp committed:
Add settings, autoload them, and check dependency

cliffordp committed:
Add tribe_notice() for Hello World

Rename main plugin file - so we can actually activate the example extension to ensure it's working.

cliffordp committed:
Start with an empty .pot file

cliffordp committed:
rename .pot file for new plugin name

barryhughes commented on src/Settings.php:
```diff
@@ -0,0 +1,87 @@
+
+namespace Tribe\Extensions\Example;
+
+if ( ! class_exists( 'Settings' ) ) {
```
Needs to be namespace aware.

cliffordp replied:
missed this comment before. done.

barryhughes commented on src/Settings.php:
```diff
@@ -0,0 +1,87 @@
+		}
+
+		/**
+		 * Removes the settings from Events > Settings > General tab > "Map Settings" section
```
Guessing this is here as an example, but can we make that clear by changing the comment to something like _"Here is an example of removing settings from..."_

cliffordp replied:
done

barryhughes commented on src/Settings.php:
```diff
@@ -0,0 +1,87 @@
+		 *
+		 * @var Settings_Helper
+		 */
+		protected $settings_helper = '';
```
No need to assign an empty string.

cliffordp replied:
👍

barryhughes commented on src/Settings_Helper.php:
```diff
@@ -0,0 +1,171 @@
+
+use Tribe__Main;
+
+if ( ! class_exists( 'Settings_Helper' ) ) {
```
Needs to be namespace aware.

cliffordp replied:
Good catch. Added to main extension file's check as well.

barryhughes reviewed (changes requested):
Thanks for the work on this!

One thing I'm less sure about is the inclusion of the Settings/Settings Helper classes. I think this is probably something where we ought to work changes back into Tribe Common. Could be worth talking that out in some more detail.

cliffordp committed:
namespace-aware class_exists()

cliffordp committed:
Make the requested changes to Settings class

cliffordp committed:
Consistency with plugin header (find and replace)

cliffordp committed:
Add all main plugin classes and minimum versions

cliffordp committed:
Fix italic placement

cliffordp committed:
Add README.md instructions, as requested by Andras

cliffordp commented:
Added some additional enhancements, particularly `README.md`
Please re-review.
Goal is to help quick-start newbies (our own or community) and to be able to remove the https://github.com/mt-support/extension-development-guide repo so we can maintain the template in a single location.

cliffordp committed:
Clarify language about Extension class existing/loading

barryhughes commented on README.md:
```diff
@@ -0,0 +1,48 @@
+    1. `[Extension Description]` with your extension plugin's short description.
+    1. PHP namespace: `Tribe\Extensions\Example`
+    1. Extension class (must be unique among all extensions): `Tribe__Extension__Example`
+    1. 
```
Delete

cliffordp replied:
good catch on the empty line

barryhughes commented on README.md:
```diff
@@ -0,0 +1,48 @@
+        1. `[Extension Name]` should be the name of your extension, such as `Remove Export Links`
+    1. `[Extension Description]` with your extension plugin's short description.
+    1. PHP namespace: `Tribe\Extensions\Example`
+    1. Extension class (must be unique among all extensions): `Tribe__Extension__Example`
```
If we're encouraging PHP namespaces I'm not sure we also need double underscore pseudo-namespacing for the class name. Is this a limitation of the existing extension framework, or can we further simplify?

cliffordp replied:
As discussed via Slack, went the route of custom namespace with `Main` class name

barryhughes commented on README.md:
```diff
@@ -0,0 +1,48 @@
+1. Double-check your plugin's slug, name, and descriptions still apply to the end result of what your code actually accomplishes.
+1. Set the `== Changelog ==` section of `readme.txt`.
+1. Delete this `README.md` file.
+1. Zip and release your plugin!
```
I love that we started with a step to _"Create a new branch called `initial` or `v1` or whatever is appropriate"_ as it helps us set things up in a way that will make pull requests/initial code review easier, but it seems like we get to _"Zip and release your plugin!"_ without an actual code review step.

Especially for new extension authors, that's important — can we add it?

cliffordp replied:
As discussed via Slack, this README.md may be useful internally but is primarily for community usage.

barryhughes commented on src/Settings_Helper.php:
```diff
@@ -4,7 +4,7 @@
 use Tribe__Main;
 
-if ( ! class_exists( 'Settings_Helper' ) ) {
+if ( ! class_exists( NS . 'Settings_Helper' ) ) {
```
:thinking: I'm not opposed to keeping these, but in a world where we encourage and give an example of autoloading, they are pretty unnecessary.

cliffordp replied:
gotcha

cliffordp committed:
Fix errant extra item

cliffordp committed:
Correct text domain

cliffordp committed:
Correct text domain

cliffordp committed:
Remove unnecessary class_exists() checks

Unnecessary since they'd only be loaded via Autoloader anyway, which won't try to load it a second time if already loaded

cliffordp committed:
Unique namespace with simplified class name
